records/2022-04-14.secrets-usage.md

```diff
@@ -0,0 +1,136 @@
+# Secrets in Notify
+
+Date: 2022-04-14
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+With potential secrets leaked to the public, we need exhaustive documentation that
+provides a quick reference on how secrets in Notify are used. This document aims to
+explain how these interact with different components of our application.
+
+## Methods and libraries
+
+This section enumerates a few methods and libraries to help understand the rest of
+the document to casual readers.
+
+### [itsdangerous](https://itsdangerous.palletsprojects.com/)
+
+This is a library that specializes in signing content in a cryptographic manner using
+a [HMAC-SHA1 algorithm](https://en.wikipedia.org/wiki/HMAC). This will not encrypt a
+message in a one-way manner; we can reverse the operation if we have the secret and
+salt used throughout the signing process. Rather, the goal is to guarantee
+authenticity and integrity of a message.
+
+> In cryptography, an HMAC (sometimes expanded as either keyed-hash message authentication code or hash-based message authentication code) is a specific type of message authentication code (MAC) involving a cryptographic hash function and a secret cryptographic key. As with any MAC, it may be used to simultaneously verify both the data integrity and authenticity of a message.
+
+> HMAC can provide authentication using a shared secret instead of using digital signatures with asymmetric cryptography. It trades off the need for a complex public key infrastructure by delegating the key exchange to the communicating parties, who are responsible for establishing and using a trusted channel to agree on the key prior to communication.
+
+> [Wikipedia](https://en.wikipedia.org/wiki/HMAC)
+
+### [JSON Web Token (JWT)](https://en.wikipedia.org/wiki/JSON_Web_Token)
+
+Notify supports JWT authentication for two scenarios: for authenticating the admin
```

Comment:
This is a little unclear for me. Does this mean that admin type users **must** use JWT and cannot the auth header with api key?

Reply:
Yes that's the case, but admin user in that case means for HTTP requests moving from the admin to the API. It's not for admin users in the UI. Does that make more sense? And would you like me to reformulate the content in the ADR? Feel free to suggest a better formulation. 🙏

Reply:
Ah, understood.  There is a little ambiguity with the term admin for me - it sounded like a user role and not the web app. 

Maybe we could refer to the "Notify admin" as the "Notify admin interface" to be more clear?

> Notify supports JWT authentication for clients accessing the API programmatically as one of two options.  The Notify admin interface uses JWT when communicating with the Notify API.

Reply:
Changed!